sklearn/cluster/_bisect_k_means.py

```diff
@@ -0,0 +1,729 @@
+                data, weights, center, label, self._n_threads
+            )
+
+        return errors_by_label
+
+    def _check_params(self, X):
+        super()._check_params(X)
+
+        # bisect_strategy
+        if self.bisect_strategy not in ["biggest_sse", "largest_cluster"]:
+            raise ValueError(
+                "Bisect Strategy must be 'biggest_sse', "
+                "or 'largest_cluster' "
+                f"got {self.bisect_strategy} instead."
+            )
+
+        # Regular K-Means should do less computations when there are only
+        # less than 3 clusters
+        if self.n_clusters < 3:
+            warnings.warn(
+                "BisectingKMeans might be inefficient for n_cluster "
+                "smaller than 3  "
+                "- Use Normal KMeans from sklearn.cluster instead.",
+                EfficiencyWarning,
+            )
+
+        if X.shape[0] <= 1:
+            raise ValueError(
+                "BisectingKMeans needs more than one sample to perform bisection."
+            )
+
+        if hasattr(self.init, "__array__"):
+            raise ValueError("BisectingKMeans does not support init as array.")
+
+    def _warn_mkl_vcomp(self, n_active_threads):
+        """Warn when vcomp and mkl are both present"""
+        warnings.warn(
+            "BisectingKMeans is known to have a memory leak on Windows "
+            "with MKL, when there are less chunks than available "
+            "threads. You can avoid it by setting the environment"
+            f" variable OMP_NUM_THREADS={n_active_threads}."
+        )
+
+    def _bisect(self, X, sample_weight, random_state):
+        """Bisection of data.
+
+        Attempts to get best bisection of data by performing regular K-Means
+        for different pairs of centroids.
+
+        .. note:: Number of attempts is specified by self.n_init
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            Training instances to cluster.
+
+            .. note:: The data will be converted to C ordering,
+                which will cause a memory copy
+                if the given data is not C-contiguous.
+
+        sample_weight : array-like of shape (n_samples,)
+            The weights for each observation in X. If None, all observations
+            are assigned equal weight.
+
+        random_state : int, RandomState instance
+            Determines random number generation for centroid initialization.
+
+        Returns
+        -------
+        self
+            Fitted estimator.
+        """
+        init = self.init
+        x_squared_norms = row_norms(X, squared=True)
+
+        best_inertia = None
+
+        for i in range(self.n_init):
+            centers_init = self._init_centroids(
+                X, x_squared_norms, init, random_state, n_centroids=2
+            )
+
+            labels, inertia, centers, _ = self._kmeans_single(
```

Comment:
Maybe I'm wrong, but I'd say: not exactly. It is used to minimalize randomness in centroids init.
But i'll give a closer look on that

Reply:
I think in most situations KMeans with 2 centers will provide quite consistent results with `init="random"` and `n_init=1`.
Keeping n_init=10 (default) and init="k-means++" (default), makes BisectingKMeans significantly slower.
They can be confusing to the users as well since they means something different than in KMeans (they only apply to the small kmeans on bisections).
My feeling would be to not expose them at all.

I'd still be curious if to see if there are some datasets for which the quality of BisectingKMeans is really improved when optimizing the init for each bisection.

Reply:
You example above is quite simple, and the situation might be different when the data is spread into more clusters or if the dimension is higher. But I get your point and I agree that 10 initialization is too much, and that even 1 init could be sufficient.

What do you think about the benefit of "kmeans++" with respects to lowering the number of iteration before convergence? Could it still be faster than "random" ?

Reply:
Being honest I had the problem with picking `n_init` for next iterations, and considered it as simplest solution.

Some Interesting things happened as I've run similar test as second one with `BisectKmeans` with `n_init=1`, but also `max_iter=1000` and `n_clusters=3`:

1. Setting in next bisections `n_init=1` ended with failed tests as cluster centers where slightly shifted (e.g [10, 3] instead of [10, 2]). 
For `n_init=1` in all bisections doesn't produce any errors

2. switching `init`:
- `init="k-means++"` : mean~=7434.512568, std~=42.329826, cv~=0.0056936921
- `init="random"`: &emsp;&ensp; mean~=7428.760366, std~=31.353423, cv~=**0.0042205458**

dummy speed test using `time.time()` in my case shows that `init="random"` seems to be 4 times faster than init="k-means++"`

Reply:
> You example above is quite simple, and the situation might be different when the data is spread into more clusters.

My point is data will never be spread into more clusters. At each step you take the data of one cluster (ignoring all other clusters and data points) and split this one into 2.

Reply:
Thanks for the experiments @michalkrawczyk . I also tested the 2 init and found similar results.

I think it's ok to keep the ``n_init`` and ``init`` parameters but I'd be in favor of setting the default to ``n_init=1`` and ``ìnit="random"``. We can keep the default "k-means++" but I'm really not convinced it's beneficial for BisectingKMeans.

Reply:
Seems reasonable. I'll change that with next commits

Reply:
@jeremiedbb I've changed to `init="random"` and `n_init=1`, but I found the catch, when checking `random_state`!=0
With `n_init=1` it's possible to achieve sligtly shifted clusters (e.g [10, 3] instead of [10, 2]). 
Also may switched clusters may occur (e.g [[10, 8], [10, 2]] instead [[10, 2], [10, 8]]), with higher probability when `init="random"`

I think it should be at least covered in docs as `.. warning::` - what do you think?

Reply:
> With n_init=1 it's possible to achieve sligtly shifted clusters (e.g [10, 3] instead of [10, 2]).

With more inits it would also be possible, just less likely. BisectingKMeans (like KMeans) does not find the absolute optimal clustering. The problem is non-convex and both algorithms will find a local optimum. ``n_inits`` helps to find better local optima but does not guarantee to find the global optimum. So it's fine that all seeds doen't give the same result, even on toy dataset.